autoPyTorch/api/base_task.py

```diff
@@ -551,89 +551,142 @@ def _do_dummy_prediction(self, num_run: int) -> None:
+        dask_futures = []
+
+        total_number_classifiers = len(available_classifiers) + num_run
         for n_r, classifier in enumerate(available_classifiers, start=num_run):
+
+            # Only launch a task if there is time
             start_time = time.time()
-            scenario_mock = unittest.mock.Mock()
-            scenario_mock.wallclock_limit = time_for_traditional_classifier_sec
-            # This stats object is a hack - maybe the SMAC stats object should
-            # already be generated here!
-            stats = Stats(scenario_mock)
-            stats.start_timing()
-            ta = ExecuteTaFuncWithQueue(
-                backend=self._backend,
-                seed=self.seed,
-                metric=self._metric,
-                logger_port=self._logger_port,
-                cost_for_crash=get_cost_of_crash(self._metric),
-                abort_on_first_run_crash=False,
-                initial_num_run=num_run,
-                stats=stats,
-                memory_limit=memory_limit,
-                disable_file_output=True if len(self._disable_file_output) > 0 else False,
-                all_supported_metrics=self._all_supported_metrics
-            )
-            dask_futures.append((classifier, self._dask_client.submit(ta.run, config=classifier,
-                                                                      cutoff=time_for_traditional_classifier_sec)))
+            if time_left >= func_eval_time_limit:
+                self._logger.info(f"{n_r}: Started fitting {classifier} with cutoff={func_eval_time_limit}")
+                scenario_mock = unittest.mock.Mock()
+                scenario_mock.wallclock_limit = time_left
+                # This stats object is a hack - maybe the SMAC stats object should
+                # already be generated here!
+                stats = Stats(scenario_mock)
+                stats.start_timing()
+                ta = ExecuteTaFuncWithQueue(
+                    backend=self._backend,
+                    seed=self.seed,
+                    metric=self._metric,
+                    logger_port=self._logger_port,
+                    cost_for_crash=get_cost_of_crash(self._metric),
+                    abort_on_first_run_crash=False,
+                    initial_num_run=n_r,
+                    stats=stats,
+                    memory_limit=memory_limit,
+                    disable_file_output=True if len(self._disable_file_output) > 0 else False,
+                    all_supported_metrics=self._all_supported_metrics
+                )
+                dask_futures.append([
+                    classifier,
+                    self._dask_client.submit(
+                        ta.run, config=classifier,
+                        cutoff=func_eval_time_limit,
+                    )
+                ])
+
+                # Increment the launched job index
+                num_run = n_r
+
+            if len(dask_futures) >= self.n_jobs:
```

Comment:
Could you explain how this portion is waiting for resources? I did not quite understand this part of the code

Reply:
thanks for the feedback, I will add a comment explaining this. Let me elaborate here before the push. Will something like this works?
```
When managing time, we need to take into account the allocated time resources,
which are dependent on the number of cores. 'dask_futures' is a proxy to the number 
of workers /n_jobs that we have, in that if there are 4 cores allocated, we can run at most
4 task in parallel. Every 'cutoff' seconds, we generate up to 4 tasks.
If we only have 4 workers and there are 4 futures in dask_futures, it means that every
worker has a task. We would not like to launch another job until a worker is available. To this
end, the following if-statement queries the number of active jobs, and forces to wait for a job
completion via future.result(), so that a new worker is available for the next iteration. 
```

Reply:
Yes, this works. Thanks for the explanation.